src/Lecture2.hs

```diff
@@ -74,7 +86,16 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt _ [] = (Nothing, [])
+removeAt n (x:xs)
+  | n < 0 = (Nothing , x:xs)
+  | n == 0 = (Just x, xs)
+  | otherwise = mergeResultTuple (Nothing , [x]) (removeAt (n -1) xs)
+
+mergeResultTuple :: (Maybe a, [a]) -> (Maybe a, [a]) -> (Maybe a, [a])
+mergeResultTuple (Nothing, l1) (Just a, l2) = (Just a, l1 ++ l2)
+mergeResultTuple (Nothing, l1) (Nothing, l2) = (Nothing, l1 ++ l2)
```

Comment:
You always provide `Nothing` in the left argument to `mergeResultTuple`. And in patterns you handle only `Nothing` for the left argument. So this is redundant and you can just remove `Maybe` from arguments. Also, you can notice that you always return the value of type `Maybe` and concatenate lists. Moreover, the list on the left side has always size `1` so instead of of using `++` you can use `:`.

After all these simplifications, you can simplify the last case to something like this:

```suggestion
  | otherwise =
      let (val, rest) = removeAt (n - 1) xs
      in (val, x : rest)
```